roll/__init__.py

```diff
@@ -301,33 +310,138 @@ def cookies(self):
         return self._cookies
 
 
-class Protocol(asyncio.Protocol):
+class WSProtocol(WebSocketCommonProtocol):
+
+    needs_upgrade = True
+    allowed_methods = {'GET'}
+    timeout = 5
+    max_size = 2 ** 20  # 1 megabytes
+    max_queue = 64
+    read_limit = 2 ** 16
+    write_limit = 2 ** 16
+
+    def __init__(self, request):
+        self.request = request
+        super().__init__(
+            timeout=self.timeout,
+            max_size=self.max_size,
+            max_queue=self.max_queue,
+            read_limit=self.read_limit,
+            write_limit=self.write_limit)
+
+    def handshake(self, response):
+        """Websocket handshake, handled by `websockets`
+        """
+        def get_header(k):
+            return self.request.headers.get(k.upper(), '')
+
+        def set_header(k, v):
+            response.headers[k] = v
+
+        try:
+            key = handshake.check_request(get_header)
+            handshake.build_response(set_header, key)
+        except InvalidHandshake:
+            raise RuntimeError('Invalid websocket request')
+
+        subprotocol = None
+        ws_protocol = get_header('Sec-Websocket-Protocol')
+        subprotocols = self.request.route.payload.get('subprotocols')
+        if subprotocols and ws_protocol:
+            # select a subprotocol
+            client_subprotocols = tuple(
+                (p.strip() for p in ws_protocol.split(',')))
+            for p in client_subprotocols:
+                if p in subprotocols:
+                    subprotocol = p
+                    set_header('Sec-Websocket-Protocol', subprotocol)
+                    break
+
+        # Return the subprotocol agreed upon, if any
+        self.subprotocol = subprotocol
+
+    async def run(self):
+        try:
+            self.request.app.websockets.add(self)
```

Comment:
I think I'd prefer let this to the user responsibility if they need it.
Mainly because I feel like this is not generic enough to be in the core (a real usecase would not combine all websockets in once, but maybe only some of them?).
We can add an how-to.
Something like:

```python3
@app.listen('startup')
async def on_startup():
    app['chat_websockets'] = set()

async def myhandler(request, ws):
    app['chat_websockets'].add(ws)
    ...
```
And writting this I see we certainly needs to expose a way to discard. Two new hooks on websocket add and discard?